airflow/decorators/assets.py

```diff
@@ -0,0 +1,151 @@
+
+if TYPE_CHECKING:
+    from typing import Sequence
+
+    from airflow.io.path import ObjectStoragePath
+
+
+class _AssetMainOperator(PythonOperator):
+    def __init__(self, *, definition_name: str, uri: str | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._definition_name = definition_name
+        self._uri = uri
+
+    def _iter_kwargs(
+        self, context: Mapping[str, Any], active_assets: dict[str, Asset]
+    ) -> Iterator[tuple[str, Any]]:
+        value: Any
+        for key in inspect.signature(self.python_callable).parameters:
+            if key == "self":
+                key = "_self"
+                value = active_assets.get(self._definition_name)
+                if not value:
+                    value = Asset(name=self._definition_name)
+                    if self._uri is not None:
+                        value.uri = self._uri
+            elif key == "context":
+                value = context
+            else:
+                value = active_assets.get(key, Asset(name=key))
+            yield key, value
+
+    def determine_kwargs(self, context: Mapping[str, Any]) -> Mapping[str, Any]:
+        active_assets: dict[str, Asset] = {}
+        asset_names = [asset_ref.name for asset_ref in self.inlets if isinstance(asset_ref, AssetRef)]
+        if "self" in inspect.signature(self.python_callable).parameters:
+            asset_names.append(self._definition_name)
+
+        if asset_names:
+            with create_session() as session:
+                active_assets = _fetch_active_assets_by_name(asset_names, session)
+        return dict(self._iter_kwargs(context, active_assets))
+
+
+def _handle_self_argument(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(_self: Any, *args: Sequence[Any], **kwargs: dict[str, Any]) -> Any:
+        return func(_self, *args, **kwargs)
+
+    return wrapper
+
+
+@attrs.define(kw_only=True)
+class AssetDefinition(Asset):
+    """
+    Asset representation from decorating a function with ``@asset``.
+
+    :meta private:
+    """
+
+    function: Callable
+    schedule: ScheduleArg
+
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        if "self" in parameters:
+            self.function: Callable = _handle_self_argument(self.function)
+
+        with DAG(dag_id=self.name, schedule=self.schedule, auto_register=True) as dag:
+            _AssetMainOperator(
+                task_id="__main__",
+                inlets=[
+                    AssetRef(name=inlet_asset_name)
+                    for inlet_asset_name in parameters
+                    if inlet_asset_name not in ("self", "context")
+                ],
+                outlets=[self.to_asset()],
+                python_callable=self.function,
+                definition_name=self.name,
+                uri=self.uri,
+            )
+
+        DAG.bulk_write_to_db([dag])
```

Comment:
Should we call this here? We should probably delegate to DagBag to do this writing instead. That’s where “normal” DAGs are written.

Reply:
Maybe this is the cause to the this? https://github.com/apache/airflow/pull/41325#discussion_r1832445526

Reply:
Yep, this is what I suspect as well 🤔 Will try it after the task_sdk pr

Reply:
removed and seems to work well